lerobot/common/policies/vqbet/modeling_vqbet.py

```diff
@@ -0,0 +1,2382 @@
+        centers = self.vqvae_model.draw_code_forward(sampled_centers).view(
+            NT, -1, self.config.vqvae_embedding_dim
+        )
+        return_decoder_input = einops.rearrange(
+            centers.clone().detach(), "NT 1 D -> NT D"
+        )
+        # pass the centroids through decoder to get actions.
+        decoded_action = (
+            self.vqvae_model.get_action_from_latent(return_decoder_input)
+            .clone()
+            .detach()
+        )
+        # reshaped extracted offset to match with decoded centroids
+        sampled_offsets = einops.rearrange(
+            sampled_offsets, "NT (W A) -> NT W A", W=self.config.n_action_pred_chunk
+        )
+        # add offset and decoded centroids
+        predicted_action = decoded_action + sampled_offsets
+        predicted_action = einops.rearrange(
+            predicted_action,
+            "(N T) W A -> N T (W A)",
+            N=N,
+            T=T,
+            W=self.config.n_action_pred_chunk,
+        )
+
+        return {
+            "cbet_logits": cbet_logits if "cbet_logits" in locals() else None,
+            "predicted_action": predicted_action,
+            "sampled_centers": sampled_centers,
+            "decoded_action": decoded_action, 
+        }
+
+    def loss_fn(self, pred, target, **kwargs):
+        """
+        for given ground truth action values (target), and prediction (pred) this function calculates the overall loss.
+
+        predicted_action: predicted action chunk (offset + decoded centroids)
+        sampled_centers: sampled centroids (code of RVQ)
+        decoded_action: decoded action, which is produced by passing sampled_centers through RVQ decoder
+        NT: batch size * T
+        T: number of action query tokens, which are process through same GPT
+        cbet_logits: probability of all codes in each layer
+        """
+        action_seq = target
+        predicted_action = pred["predicted_action"]
+        sampled_centers = pred["sampled_centers"]
+        decoded_action = pred["decoded_action"]
+        NT = predicted_action.shape[0] * predicted_action.shape[1]
+        T = predicted_action.shape[1]
+        cbet_logits = pred["cbet_logits"]
+
+        predicted_action = einops.rearrange(
+            predicted_action, "N T (W A) -> (N T) W A", W=self.config.n_action_pred_chunk
+        )
+
+        action_seq = einops.rearrange(action_seq, "N T W A -> (N T) W A")
+        # Figure out the loss for the actions.
+        # First, we need to find the closest cluster center for each ground truth action.
+        state_vq, action_bins = self.vqvae_model.get_code(
+            action_seq
+        )  # action_bins: NT, G
+
+        # Now we can compute the loss.
+        if action_seq.ndim == 2:
+            action_seq = action_seq.unsqueeze(0)
+
+        # offset loss is L1 distance between the predicted action and ground truth action
+        offset_loss = torch.nn.L1Loss()(action_seq, predicted_action)
```

Comment:
nit: Use functional rather than instantiating a module:

```suggestion
        offset_loss = F.l1_loss(action_seq, predicted_action)
```

Reply:
fixed